Cleanup contributing (#1384)
thin-edge/thin-edge.io: CONTRIBUTING.md

reasonable. Most important is to reference the GitHub issue that you are
addressing with the PR.

**NOTE:** GitHub has [a feature](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
that will automatically close issues referenced with a keyword (such as "Fixes")
by a PR or commit once the PR/commit is merged. Don't use these keywords. We
don't want issues to be automatically closed. We want our testers to

> independently verify and close them.

Author: Okay, I will push a revert for this! :+1:

> independently verify and close them.

Reviewer: Thank you,

reasonable. Most important is to reference the GitHub issue that you are
addressing with the PR.

## Writing good commit messages

Having clear, concise and well-written commit messages for all changes is not

> only a indicator for the seriousness of the project, but also a way to ensure

Reviewer 2: ```suggestion
only an indicator for the seriousness of the project, but also a way to ensure
```

sustainable development via replicability of the sources of the project.

Because of this, we try to ensure the highest possible quality for each
individual commit message.

Because such a thing cannot necessarily or in full be enforced via tooling, it
remains mainly the obligation of a reviewer of a change-set to ensure

* Atomicity of the individual commits
* that one commit is one atomic change, and not an accumulation of several
individual and separate changes
* that the commit message of the change expresses why the change was made

Every reviewer of a pull request is asked to review not only the changes, but
also the commit messages.
Reviewers are explicitly empowered and actually encouraged to block pull
requests that contain commit messages that do not meet a certain standard
even and especially also if the proposed changes are acknowledged.

> The contributor should document why they implemented a change properly. A good

Reviewer 2: ```suggestion
The contributor should document why they implemented a change. A good
```

rule of thumb is that if a change is larger than 60 lines, a commit message
without a body won't suffice anymore.
The more lines are touched by an individual commit, the more lines should be
used to explain the change.

> The project implements lints using github actions to meet such criteria.

Reviewer 2: ```suggestion
```

> Also, some hard criteria is checked via github action lints:

Reviewer 2: ```suggestion
Also, some hard criteria are checked via github action lints:
```

* The `Signed-off-by` trailer lint must be present in all commits (also see
section "Ensuring the Signed-off-by-trailer")
* The commit body must meet the default formatting of a commit message, that
is

> * Subject line not longer than 50 characters
> * Subject capitalized
> * No period at the end of the subject
> * Imperative mood in the subject
> * Second line empty
> * Body lines not longer than 72 characters

Reviewer 2: ```suggestion
    * Subject line not longer than 50 characters
    * Subject capitalized
    * No period at the end of the subject
    * Imperative mood in the subject
    * Second line empty
    * Body lines not longer than 72 characters
```

Commits using the `--fixup` or `--squash` feature of `git-commit` are allowed,

> but won't be merged (read more about this in the "merge strategies" section).

Reviewer 2: ```suggestion
but will have to be squashed into the history before being merged.
Read more about this in the "merge strategies" section.
```

Author: ```suggestion
but will have to be squashed by the pull request author before the being merged.
```

Reviewer 2: remove 'the' and that's good

### Ensuring the Signed-off-by-trailer

Contributors need to sign the "Contributor License Agreement". They do so by
signing off each individual commit with a so-called "Signed-off-by Trailer".
Git offers the `--signoff`/`-s` flags when using `git-commit` to commit
changes.

To ensure that all commits have the trailer present in the commit message, a
CI lint is installed in the github actions workflows of the project. This lint

> blocks pull requests to be merged if one or more of the commits of the pull

Reviewer 2: ```suggestion
blocks merging pull requests if one or more of the commits of the pull
```

or 

```suggestion
blocks pull requests from being merged if one or more of the commits of the pull
```

## Writing good commit messages

Git commit messages should explain the how and why of your change and be
separated into a brief subject line followed by a more detailed body. When in
doubt, follow this guide for good commit messages and you can’t go wrong:

> [https://chris.beams.io/posts/git-commit/](https://chris.beams.io/posts/git-commit/).

Author: Feel free to resolve this if you think [the fixup](https://github.com/thin-edge/thin-edge.io/pull/1384/commits/93056a070969d8a1f47a363ec35f62aaf5782a0f) is okay.

* that the commit message of the change expresses why the change was made

Every reviewer of a pull request is asked to review not only the changes, but
also the commit messages.
Reviewers are explicitly empowered and actually encouraged to block pull
requests that contain commit messages that do not meet a certain standard
even and especially also if the proposed changes are acknowledged.

The contributor should document why they implemented a change. A good
rule of thumb is that if a change is larger than 60 lines, a commit message
without a body won't suffice anymore.
The more lines are touched by an individual commit, the more lines should be
used to explain the change.

Also, some hard criteria are checked via github action lints:

* The `Signed-off-by` trailer lint must be present in all commits (also see
section "Ensuring the Signed-off-by-trailer")
* The commit body must meet the default formatting of a commit message, that
is
    * Subject line not longer than 50 characters
    * Subject capitalized
    * No period at the end of the subject
    * Imperative mood in the subject
    * Second line empty
    * Body lines not longer than 72 characters

> You can also read about commit message in

Reviewer 2: ```suggestion
You can also read about commit messages in
```

master" but also to streamline the workflow for integrating pull requests.

The following actions are **not** allowed:

* Committing to `master`/`main` directly.
  The GitHub repository is configured to block pushing to `master`/`main`.
* Squash-Merging of pull requests (which is equivalent to committing to
  `master`/`main` directly).
  The github repository is configured to not allow squash merges.
* Merging of "fixup!" or "squash!" commits. A github actions
  job is employed to block pull requests from being merged if such commits are
  in the branch of the pull request. Rebase should be used to clean those up.

It is explicitly _not_ forbidden to have merge-commits in a pull request.
Long-running pull requests that contain a large number of changes and are
developed over a long period might contain merges. They might even merge
`master`/`main` to get up-to-date with the latest developments. Though it is
not encouraged to have such long-running pull requests and discouraged to
merge `master`/`main` into a PR branch, the project acknowledges that
sometimes it is not avoidable.

To summarize, the following requirements have to be met before a pull request
can be merged:

> * Reviews of the relevant persons (ensured via github setting)

Reviewer 2: ```suggestion
* Reviews of the relevant people (ensured via github setting)
```
